### PR TITLE
externalgettup_custom() ereports WARNING instead of ERROR if

### PR DESCRIPTION
**Pull request by fanfuxiaoran**
external data file is incomplete.

### CODE REVIEW

**Comment by danielgustafsson** [comment]
Is it possible to add a testcase for this?

**Comment by fanfuxiaoran** [comment]
Need a custom formatter to test this. We have custom formatters and test
cases in gpss project. Yeah, it is better to have a test case in gpdb. I
will try to add a testcase for it in gpdb.

**Comment by danielgustafsson** [comment]
> On 3 Aug 2018, at 13:42, Xiaoran Wang <notifications@github.com> wrote:
> 
> Need a custom formatter to test this. We have custom formatters and test
> cases in gpss project. Yeah, it is better to have a test case in gpdb. I
> will try to add a testcase for it in gpdb.

Thanks, that would be great!

File: src/backend/access/external/fileam.c [hunk 1]
```diff
@@ -961,139 +961,135 @@ externalgettup_custom(FileScanDesc scan)
...
+						scan->in_functions,
+						scan->typioparams);
+				(void) FunctionCallInvoke(&fcinfo);
+
+			}
+			PG_CATCH();
+			{
+				error_caught = true;
+
+				MemoryContextSwitchTo(formatter->fmt_perrow_ctx);
```
Comment (asimrp) [inline src/backend/access/external/fileam.c]: Don't you want to save the old context?

File: src/backend/access/external/fileam.c [hunk 2]
```diff
@@ -961,139 +961,135 @@ externalgettup_custom(FileScanDesc scan)
...
-				{
-					switch (formatter->fmt_notification)
-					{
-						case FMT_NONE:
+				FILEAM_HANDLE_ERROR;
 
-							/* got a tuple back */
+				MemoryContextSwitchTo(oldctxt);
+			}
+			PG_END_TRY();
```
Comment (asimrp) [inline src/backend/access/external/fileam.c]: A call to `FlushErrorState()` may be necessary in the `PG_CATCH()` block.

File: src/backend/access/external/fileam.c (hunk 1 above)
Comment (fanfuxiaoran) [reply src/backend/access/external/fileam.c]: The old context has be saved at the beginning of the function.

**Pull request merged**